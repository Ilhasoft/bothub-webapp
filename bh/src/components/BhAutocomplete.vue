<template>
  <bh-input>
    <div :class="inputClassAttr">
      <div
        v-if="$slots.prepend"
        :class="`${className}__prepend`"
      >
        <div :class="`${className}__prepend__content`">
          <slot name="prepend" />
        </div>
      </div>
      <input
        id="autocomplete"
        ref="input"
        v-model="val"
        autocomplete="off"
        :class="`${className}__input`"
        v-bind="$attrs"
        type="text"
        @keydown.38="keyArrows('up')"
        @keydown.40="keyArrows('down')"
        @keydown.13="enterPressed()"
        @focus="onFocus()"
        @blur="onBlur()"
        @keyup.esc="onEsc()"
      >
      <slot name="append" />
    </div>
    <div
      v-if="autocompleteOpen"
      ref="dropdownAutocomplete"
      class="bh-autocomplete__list"
    >
      <button
        v-for="(item, i) in filteredData"
        :key="i"
        type="button"
        :class="{ 'bh-autocomplete__list__item__active': item === hovered,
                  'bh-autocomplete__list__item': true }"
        @click="select(item)"
        v-html="highlightVerboseItem(item)"
      />
    </div>
  </bh-input>
</template>

<script>
import BhText from './BhText';


const strTrueIndexOf = (a, b) => (a
  .toString()
  .toLowerCase()
  .indexOf(b.toLowerCase())
);

export default {
  name: 'BhAutocomplete',
  extends: BhText,
  props: {
    data: {
      type: Array,
      required: true,
    },
    verboseField: {
      type: String,
      default: null,
    },
    indexField: {
      type: String,
      default: null,
    },
    forceSelectFirst: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      className: 'bh-autocomplete',
      selected: null,
      focused: false,
      hovered: null,
      openOnFocus: Boolean,
      closeForced: false,
    };
  },
  computed: {
    filteredData() {
      return this.val
        ? this.data
          .filter(item => strTrueIndexOf(this.verboseItem(item), this.val) >= 0)
          .sort((a, b) => {
            const aIndexOf = strTrueIndexOf(this.verboseItem(a), this.val);
            const bIndexOf = strTrueIndexOf(this.verboseItem(b), this.val);
            return aIndexOf - bIndexOf;
          })
        : this.data;
    },
    autocompleteOpen() {
      if (this.closeForced) {
        return false;
      }

      if (this.filteredData.length === 0) {
        return false;
      }

      if (this.focused) {
        return true;
      }

      return false;
    },
  },
  watch: {
    selected(item) {
      if (item) {
        this.$emit('selected', this.indexItem(item));
      } else {
        this.$emit('selected', null);
      }
    },
    val() {
      this.closeForced = false;
    },
  },
  methods: {
    verboseItem(item) {
      return this.verboseField
        ? item[this.verboseField]
        : item;
    },
    highlightVerboseItem(item) {
      const regex = new RegExp(`(${this.val})`, 'gi');
      return this.verboseItem(item).replace(regex, '<b>$1</b>');
    },
    indexItem(item) {
      return this.indexField
        ? item[this.indexField]
        : item;
    },
    select(item) {
      if (item) {
        this.val = this.verboseItem(item);
        this.selected = item;
        this.setHovered(false);
      } else {
        this.val = '';
        this.selected = null;
        this.setHovered(false);
      }
    },
    enterPressed() {
      if (this.hovered === null) return;
      this.select(this.hovered);
    },
    finish() {
      if (this.forceSelectFirst) {
        if (this.val) {
          const firstItem = this.filteredData[0] || this.data[0];
          if (firstItem) {
            this.select(firstItem);
          }
        } else {
          this.select();
        }
      }
    },
    keyArrows(direction) {
      const sum = direction === 'down' ? 1 : -1;
      if (this.focused) {
        let index = this.data.indexOf(this.hovered) + sum;
        index = index > this.data.length - 1 ? this.data.length : index;
        index = index < 0 ? 0 : index;

        this.setHovered(this.data[index]);

        const list = document.querySelector('.bh-autocomplete__list');
        const element = list.querySelectorAll('.bh-autocomplete__list__item')[index];

        if (!element) return;

        const visMin = list.scrollTop;
        const visMax = list.scrollTop + list.clientHeight - element.clientHeight;

        if (element.offsetTop < visMin) {
          list.scrollTop = element.offsetTop;
          this.currentHeightlist = list.scrollTop;
        } else if (element.offsetTop >= visMax) {
          list.scrollTop = (
            element.offsetTop
                            - list.clientHeight
                            + element.clientHeight
          );
          this.currentHeightlist = list.scrollTop;
        }
      } else {
        this.focused = true;
      }
    },
    onFocus() {
      this.setFocused(true);
    },
    onBlur() {
      this.currentSelectedItemIndex = -1;
      this.finish();
      this.setFocused(false);
      this.setHovered(false);
    },
    onEsc() {
      this.currentSelectedItemIndex = -1;
      this.finish();
      this.forceClosed();
      this.setHovered(false);
    },
    setFocused(value) {
      setTimeout(() => {
        this.closeForced = false;
        this.focused = value;
      }, 200);
    },
    forceClosed() {
      this.closeForced = true;
    },
    setHovered(option) {
      if (option === undefined || false) return;

      this.hovered = option;
    },
  },
};
</script>

<style lang="scss">
@import '@scss/colors.scss';
@import '@scss/variables.scss';
@import '@scss/forms.scss';


.bh {
  &-autocomplete {
    @include text();

    &__list {
      position: absolute;
      top: calc(100% + 4px);
      left: 0;
      z-index: $autocomplete-z-index;
      width: 100%;
      max-height: 150px;
      padding: .5rem 0;
      overflow: auto;
      background-color: white;
      border: solid 1px $color-grey;
      border-radius: 4px;

      &__item {
        display: block;
        width: 100%;
        padding: .5rem;
        font-size: 1rem;
        text-align: left;
        cursor: pointer;
        background-color: transparent;
        border: none;
        outline: none;

        &__active {
          color: white;
          background-color: $color-primary;
        }

        &:hover {
          color: white;
          background-color: $color-primary;
        }
      }
    }
  }
}
</style>
