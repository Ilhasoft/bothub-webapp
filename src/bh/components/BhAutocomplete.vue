<template>
  <bh-input>
    <div :class="inputClassAttr">
      <div
        v-if="$slots.prepend"
        :class="`${className}__prepend`">
        <div :class="`${className}__prepend__content`">
          <slot name="prepend" />
        </div>
      </div>
      <input
        ref="input"
        :class="`${className}__input`"
        v-bind="$attrs"
        v-model="val"
        type="text"
        @focus="onFocus()"
        @blur="onBlur()"
        @keyup.esc="onEsc()" >
      <slot name="append"/>
    </div>
    <div
      v-if="autocompleteOpen"
      class="bh-autocomplete__list">
      <button
        v-for="(item, i) in filteredData"
        :key="i"
        type="button"
        class="bh-autocomplete__list__item"
        @click="select(item)"
        v-html="highlightVerboseItem(item)" />
    </div>
  </bh-input>
</template>

<script>
import BhText from './BhText';


const strTrueIndexOf = (a, b) => (a
  .toString()
  .toLowerCase()
  .indexOf(b.toLowerCase())
);

export default {
  name: 'BhAutocomplete',
  extends: BhText,
  props: {
    data: {
      type: Array,
      required: true,
    },
    verboseField: {
      type: String,
      default: null,
    },
    indexField: {
      type: String,
      default: null,
    },
    forceSelectFirst: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      className: 'bh-autocomplete',
      selected: null,
      focused: false,
      closeForced: false,
    };
  },
  computed: {
    filteredData() {
      return this.val
        ? this.data
          .filter(item => strTrueIndexOf(this.verboseItem(item), this.val) >= 0)
          .sort((a, b) => {
            const aIndexOf = strTrueIndexOf(this.verboseItem(a), this.val);
            const bIndexOf = strTrueIndexOf(this.verboseItem(b), this.val);
            return aIndexOf - bIndexOf;
          })
        : this.data;
    },
    autocompleteOpen() {
      if (this.closeForced) {
        return false;
      }

      if (this.filteredData.length === 0) {
        return false;
      }

      if (this.focused) {
        return true;
      }

      return false;
    },
  },
  watch: {
    selected(item) {
      if (item) {
        this.$emit('selected', this.indexItem(item));
      } else {
        this.$emit('selected', null);
      }
    },
    val() {
      this.closeForced = false;
    },
  },
  methods: {
    verboseItem(item) {
      return this.verboseField
        ? item[this.verboseField]
        : item;
    },
    highlightVerboseItem(item) {
      const regex = new RegExp(`(${this.val})`, 'gi');
      return this.verboseItem(item).replace(regex, '<strong>$1</strong>');
    },
    indexItem(item) {
      return this.indexField
        ? item[this.indexField]
        : item;
    },
    select(item) {
      if (item) {
        this.val = this.verboseItem(item);
        this.selected = item;
      } else {
        this.val = '';
        this.selected = null;
      }
    },
    finish() {
      if (this.forceSelectFirst) {
        if (this.val) {
          const firstItem = this.filteredData[0] || this.data[0];
          if (firstItem) {
            this.select(firstItem);
          }
        } else {
          this.select();
        }
      }
    },
    onFocus() {
      this.setFocused(true);
    },
    onBlur() {
      this.finish();
      this.setFocused(false);
    },
    onEsc() {
      this.finish();
      this.forceClosed();
    },
    setFocused(value) {
      setTimeout(() => {
        this.closeForced = false;
        this.focused = value;
      }, 100);
    },
    forceClosed() {
      this.closeForced = true;
    },
  },
};
</script>

<style lang="scss">
@import '~bh/assets/scss/colors.scss';
@import '~bh/assets/scss/typography.scss';
@import '~bh/assets/scss/variables.scss';
@import '~bh/assets/scss/forms.scss';


.bh {
  &-autocomplete {
    @include text();

    &__list {
      @include typography();

      position: absolute;
      top: calc(100% + 4px);
      left: 0;
      z-index: $autocomplete-z-index;
      width: 100%;
      max-height: 150px;
      padding: .5rem 0;
      overflow: auto;
      background-color: white;
      border: solid 1px $color-lighter-grey;
      border-radius: 4px;

      &__item {
        display: block;
        width: 100%;
        padding: .5rem;
        font-size: 1rem;
        text-align: left;
        cursor: pointer;
        background-color: transparent;
        border: none;
        outline: none;

        &:hover {
          color: $color-primary;
          background-color: $color-lighter-grey;
        }
      }
    }
  }
}
</style>
