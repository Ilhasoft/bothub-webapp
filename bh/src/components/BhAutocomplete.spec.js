import { shallowMount, createLocalVue } from '@vue/test-utils';

import BH from '../index';
import BhAutocomplete from './BhAutocomplete';


const localVue = createLocalVue();
localVue.use(BH);

describe('BhAutocomplete.vue', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(BhAutocomplete, {
      localVue,
      propsData: {
        data: [
          'a',
          'ab',
          'abcd',
          'b',
          'c',
        ],
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('focus on input', () => {
    beforeEach(async () => {
      const input = wrapper.find({ ref: 'input' });
      input.trigger('focus');
      await new Promise(resolve => (setTimeout(resolve, 200)));
    });

    test('autocomplete is open', () => {
      expect(wrapper.vm.autocompleteOpen).toBeTruthy();
    });

    test('renders correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    describe('press esc', () => {
      beforeEach(() => {
        const input = wrapper.find({ ref: 'input' });
        input.trigger('keyup.esc');
      });

      test('autocomplete is close', () => {
        expect(wrapper.vm.autocompleteOpen).toBeFalsy();
      });

      test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe('type "b" in input', () => {
    const value = 'b';

    beforeEach(() => {
      const input = wrapper.find({ ref: 'input' });
      input.element.value = value;
      input.trigger('input');
    });

    test(`val equal ${value}`, () => {
      expect(wrapper.vm.val).toBe(value);
    });

    test(`${value} in top of filteredData`, () => {
      expect(value).toBe(wrapper.vm.filteredData[0]);
    });

    describe('select first item', () => {
      beforeEach((() => {
        wrapper.vm.select(wrapper.vm.filteredData[0]);
      }));

      test('selected event emitted', () => {
        expect(wrapper.emitted('selected')).toBeDefined();
      });

      describe('select empty (null) item', () => {
        beforeEach((() => {
          wrapper.vm.select(null);
        }));

        test('selected event emit null value', () => {
          const eventValues = wrapper.emitted('selected');
          const eventValue = eventValues[eventValues.length - 1][0];
          expect(eventValue).toBeNull();
        });
      });
    });
  });

  describe('type "z" in input', () => {
    const value = 'z';

    beforeEach(() => {
      const input = wrapper.find({ ref: 'input' });
      input.element.value = value;
      input.trigger('input');
    });

    test(`val equal ${value}`, () => {
      expect(wrapper.vm.val).toBe(value);
    });

    test('autocomplete is closed', () => {
      expect(wrapper.vm.autocompleteOpen).toBeFalsy();
    });
  });

  describe('enable forceSelectFirst', () => {
    beforeEach(() => {
      wrapper.setProps({
        forceSelectFirst: true,
      });
    });

    describe('type "abc" in input', () => {
      beforeEach(() => {
        const input = wrapper.find({ ref: 'input' });
        input.element.value = 'abc';
        input.trigger('input');
      });

      describe('blur on input', () => {
        beforeEach(() => {
          const input = wrapper.find({ ref: 'input' });
          input.trigger('blur');
        });

        test('selected event emitted', () => {
          expect(wrapper.emitted('selected')).toBeDefined();
        });

        test('selected event emitted with "abcd" value', () => {
          expect(wrapper.emitted('selected')[0][0]).toBe('abcd');
        });
      });
    });

    describe('type "z" in input', () => {
      beforeEach(() => {
        const input = wrapper.find({ ref: 'input' });
        input.element.value = 'z';
        input.trigger('input');
      });

      describe('blur on input', () => {
        beforeEach(() => {
          const input = wrapper.find({ ref: 'input' });
          input.trigger('blur');
        });

        test('selected event emitted', () => {
          expect(wrapper.emitted('selected')).toBeDefined();
        });

        test('selected event emitted with "z" value', () => {
          expect(wrapper.emitted('selected')[0][0]).toBe('a');
        });

        describe('blank input finish', () => {
          beforeEach(() => {
            const input = wrapper.find({ ref: 'input' });
            input.element.value = '';
            input.trigger('input');
          });

          describe('blur on input', () => {
            beforeEach(() => {
              const input = wrapper.find({ ref: 'input' });
              input.trigger('blur');
            });

            test('selected event emitted', () => {
              expect(wrapper.emitted('selected')[1][0]).toBeNull();
            });
          });
        });
      });
    });
  });

  describe('using verboseField and indexField', () => {
    beforeEach(() => {
      wrapper.setProps({
        verboseField: 'label',
        indexField: 'id',
        data: [
          { id: 1, label: 'a' },
          { id: 2, label: 'b' },
        ],
      });
    });

    describe('type "b" in input', () => {
      const value = { id: 2, label: 'b' };
      const query = 'b';

      beforeEach(() => {
        const input = wrapper.find({ ref: 'input' });
        input.element.value = query;
        input.trigger('input');
      });

      test('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
      });

      test(`val equal ${query}`, () => {
        expect(wrapper.vm.val).toBe(query);
      });

      test(`${value} in top of filteredData`, () => {
        expect(value).toMatchObject(wrapper.vm.filteredData[0]);
      });

      describe('select first item', () => {
        beforeEach((() => {
          wrapper.vm.select(wrapper.vm.filteredData[0]);
        }));

        test('selected event emitted', () => {
          expect(wrapper.emitted('selected')).toBeDefined();
        });

        test('selected event emitted value\'s id', () => {
          expect(wrapper.emitted('selected')[0][0]).toBe(2);
        });
      });
    });
  });
});
