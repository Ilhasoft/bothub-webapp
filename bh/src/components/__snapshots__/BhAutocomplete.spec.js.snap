// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BhAutocomplete.vue focus on input press esc renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`BhAutocomplete.vue focus on input renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`BhAutocomplete.vue renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;

exports[`BhAutocomplete.vue using verboseField and indexField type "b" in input renders correctly 1`] = `
VueWrapper {
  "_emitted": Object {
    "input": Array [
      Array [
        "b",
      ],
    ],
  },
  "_emittedByOrder": Array [
    Object {
      "args": Array [
        "b",
      ],
      "name": "input",
    },
  ],
  "isFunctionalComponent": undefined,
}
`;
